### PR TITLE
feat: add dsa support

### PR DESCRIPTION
**Pull request by dignifiedquire**
Based on #124

Depends on https://github.com/RustCrypto/signatures/pull/798

TODOs

- [x] Implement verification
- [x] Update to released version of https://github.com/RustCrypto/signatures/pull/798
- [x] Implement key creation
- [x] Update key tests

### CODE REVIEW

**Comment by hko-s** [comment]
The interop test suite didn't yield any interesting results for this PR, so I decided to test data signatures manually, using a few SOP implementations.

This comment documents my testing procedure. I'm attaching the test artifacts.

## Generate test key

As setup, we generate a dsa3072 key with rpgp:

```
$ rsop generate-key --profile dsa alice@example.org > key.dsa
# ^ we generate a key with dsa3072 primary [CS]
```

## Generate data signature with rsop (rpgp), verify with different implementations

We then produce a data signature with rsop (rpgp):

```
$ echo -n "hello world" | rsop sign key.dsa > sig.rpgp
```

...  and verify it with other SOP implementations:

```
$ echo -n "hello world" | rsop verify sig.rpgp key.dsa
2024-02-06T22:39:11Z 13375374f22eb5a0563137673cb7f9b7a8bb79a2 13375374f22eb5a0563137673cb7f9b7a8bb79a2 mode:binary

$ echo -n "hello world" | sqop verify sig.rpgp key.dsa
2024-02-06T22:39:11Z 13375374F22EB5A0563137673CB7F9B7A8BB79A2 13375374F22EB5A0563137673CB7F9B7A8BB79A2 mode:binary

$ echo -n "hello world" | pgpainless-cli verify sig.rpgp key.dsa
2024-02-06T22:39:11Z 13375374F22EB5A0563137673CB7F9B7A8BB79A2 13375374F22EB5A0563137673CB7F9B7A8BB79A2 mode:binary

$ echo -n "hello world" | ~/go/bin/gosop verify sig.rpgp key.dsa
panic: runtime error: slice bounds out of range [:5] with length 3

goroutine 1 [running]:
github.com/ProtonMail/gosop/utils.ReadFileOrEnv({0x7ffe2b35946c?, 0xc00004e028?})
    /home/heiko/src/gosop/utils/utils.go:132 +0xeb
github.com/ProtonMail/gosop/cmd.Verify({0xc000073940, 0xadf440?, 0x411f25?})
    /home/heiko/src/gosop/cmd/verify.go:46 +0x185
github.com/ProtonMail/gosop/cmd.glob..func5(0xc000016b00?)
    /home/heiko/src/gosop/cmd/commands.go:64 +0x3c
github.com/urfave/cli/v2.(*Command).Run(0xae0660, 0xc000016ac0)
    /home/heiko/go/pkg/mod/github.com/urfave/cli/v2@v2.2.0/command.go:164 +0x583
github.com/urfave/cli/v2.(*App).RunContext(0xc000002180, {0x81b2a0?, 0xb2d140}, {0xc0000160c0, 0x4, 0x4})
    /home/heiko/go/pkg/mod/github.com/urfave/cli/v2@v2.2.0/app.go:306 +0xb16
github.com/urfave/cli/v2.(*App).Run(...)
    /home/heiko/go/pkg/mod/github.com/urfave/cli/v2@v2.2.0/app.go:215
main.main()
    /home/heiko/src/gosop/main.go:36 +0x1eb

$ ~/go/bin/gosop version --extended
gosop 0.1.0
GopenPGP 2.7.4
ProtonMail/go-crypto 5aa5874ade95
x/crypto v0.7.0
Compiled using go1.21.6
```

gosop crashes while trying to verify the signature (so far I have no idea why).

## Generate data signature with sqop (sequoia), verify with different implementations

For comparison, we create a signature with sqop (sequoia), using the same private key (generated by rpgp):

```
$ echo -n "hello world" | sqop sign key.dsa > sig.seq
```

Verifying this signature works with all SOP implementations used above, and also with gosop (which failed above):

```
$ echo -n "hello world" | ~/go/bin/gosop verify sig.seq key.dsa
2024-02-06T22:40:32Z 13375374F22EB5A0563137673CB7F9B7A8BB79A2 13375374F22EB5A0563137673CB7F9B7A8BB79A2
```

Further research into the gosop verification failure seems to be called for.

[sig.seq.txt](https://github.com/rpgp/rpgp/files/14187013/sig.seq.txt)
[key.dsa.txt](https://github.com/rpgp/rpgp/files/14187014/key.dsa.txt)
[sig.rpgp.txt](https://github.com/rpgp/rpgp/files/14187015/sig.rpgp.txt)

File: src/crypto/dsa.rs [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+    let components = Components::from_components(p, q, g)?;
+    let verifying_key = VerifyingKey::from_components(components, y)?;
+    let signature = Signature::from_components(r, s)?;
+
+    verifying_key.verify_prehash(hashed, &signature)?;
+
+    Ok(())
+}
+
+/// Generate an DSA KeyPair.
```
Comment (hko-s) [inline src/crypto/dsa.rs:60]:
```suggestion
/// Generate a DSA KeyPair.
```

File: tests/key_test.rs [hunk 2]
```diff
@@ -114,16 +114,16 @@ macro_rules! parse_dumps {
...
-    (test_parse_dumps_1, 1, 18_021, 21_000),
-    (test_parse_dumps_2, 2, 18_058, 20_999),
-    (test_parse_dumps_3, 3, 18_097, 20_998),
-    (test_parse_dumps_4, 4, 18_048, 20_999),
-    (test_parse_dumps_5, 5, 18_090, 20_999),
-    (test_parse_dumps_6, 6, 18_130, 21_000),
-    (test_parse_dumps_7, 7, 18_133, 21_000),
-    (test_parse_dumps_8, 8, 18_139, 21_000),
-    (test_parse_dumps_9, 9, 17_992, 21_000),
+    (test_parse_dumps_0, 0, 17_697, 20_998),
```
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply tests/key_test.rs:117]: so before we completely ignored dsa keys, so this number was expected to go down, I haven’t had time to dig into the failures in detail though

**Comment by hko-s** [comment]
I've concluded my review. Besides being surprised by the numbers in the `test_parse_dumps` tests, nothing stood out to me.

File: tests/key_test.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply tests/key_test.rs:117]: that is very much possible, we should investigate this, but I didn't want to block getting the basic support in by this

[the remark this answers is not in the document]
Comment (hko-s) [reply tests/key_test.rs:117]: makes sense. opening a ticket to investigate this some other time might be nice as a reminder.

[the remark this answers is not in the document]
Comment (dignifiedquire) [reply tests/key_test.rs:117]: https://github.com/rpgp/rpgp/issues/284

[the remark this answers is not in the document]
Comment (hko-s) [reply tests/key_test.rs:117]: Noting down a stray thought, here:

In rPGP's `verify` functionality, third party signatures where/are considered an "error" in TPKs (also see https://github.com/rpgp/rpgp/pull/448).

Maybe by implementing DSA support, more third party signatures moved into view of the `verify` mechanism, thus invalidating all TPKs that have third-party certifying signatures based on DSA?

[the remark this answers is not in the document]
Comment (dignifiedquire) [reply tests/key_test.rs:117]: might be, yes

**Pull request merged**